### ISSUE TITLE
Fix pre-existing memory and allocation bugs

### ISSUE THREAD

**Issue by LeoBuron** [es-ude/OnDeviceTraining#67]
## Summary

Pre-existing bugs discovered during review of #66 plan. Should be fixed **before** the training loop refactoring to keep that PR clean.

### 1. `inferenceStats_t *` memory leak in MnistExperiment.c eval loop

**File:** `experiments/MnistExperiment.c:294-299`

`inferenceWithLoss()` is called 10,000 times per epoch and the returned `inferenceStats_t *` is never freed. Each call allocates a stats struct + output tensor via `reserveMemory`, leaking ~hundreds of bytes per call × 10K × 10 epochs.

**Fix:** Add `freeInferenceStats(validationResult)` after reading the loss value.

### 2. `batch_t *` memory leak in MnistExperiment.c eval loop

**File:** `experiments/MnistExperiment.c:294-299`

Same validation loop as bug 1: `testDataloader->getBatch()` allocates a `batch_t *` via `reserveMemory` each iteration, but `freeBatch(batch)` is never called. The accuracy loop (line 314) correctly frees both the result and the batch — the validation loop does neither.

**Fix:** Add `freeBatch(batch)` after reading the loss value.

### 3. `batch->samples` allocation uses wrong sizeof

**File:** `src/userApi/data_loader/DataLoaderApi.c:64`

```c
batch->samples = *reserveMemory(batchSize * sizeof(sample_t));  // wrong
```

`batch->samples` is `sample_t **` (array of pointers), so this should be `sizeof(sample_t *)`. Currently over-allocates (sample_t is 16 bytes, pointer is 8 bytes on 64-bit), so it does not crash — but it wastes memory and is incorrect on principle.

**Fix:** Change to `sizeof(sample_t *)`.

## Context

Found during codebase exploration for #66.
